In [1]:
#Import packages:
import pandas as pd
import numpy as np
import os

In [2]:
# Create string of location
path = '/Users/aegarcia/Documents/Documents/01-2022 Instacart Basket Analysis/'
# Import products.csv
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)
# Import orders.csv
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

In [3]:
# Create a dataframe
df_test = pd.DataFrame()

# 4.5 Lesson
## Mixed data types

In [4]:
# Create a mixed type column
df_test['mix'] = ['a', 'b', 1, True]

In [5]:
df_test.head()

mix
0     a
1     b
2     1
3  True

In [6]:
# Check for mixed types
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

mix


In [7]:
df_test.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4 entries, 0 to 3
Data columns (total 1 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   mix     4 non-null      object
dtypes: object(1)
memory usage: 160.0+ bytes


In [8]:
# Change all type to string
df_test['mix'] = df_test['mix'].astype('str')

In [9]:
# Check for mixed types again, and now, we DID NOT GET MIX!
for col in df_test.columns.tolist():
  weird = (df_test[[col]].applymap(type) != df_test[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_test[weird]) > 0:
    print (col)

## Missing values

In [10]:
# Get counts of missing values per column
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [11]:
# Create a dataframe with the 16 null values
df_nan = df_prods[df_prods['product_name'].isnull() == True]
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [12]:
# Use mean or median to impute numeric nans
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

In [13]:
# The missing values are stings, lets look at the dimensions of our df
df_prods.shape

(49693, 5)

In [14]:
# Create a new df without those 16 obs
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]
# ==False becasue we want non-missing values
# or could use this instead
# df_prods[‘product_name’].dropna(inplace = True)

In [15]:
# and this gives us 16 less
df_prods_clean.shape

(49677, 5)

## Duplicates

In [16]:
# Create new df using duplicated()
df_dups = df_prods_clean[df_prods_clean.duplicated()]
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [17]:
#look for the dimension of df
df_prods_clean.shape

(49677, 5)

In [18]:
# Remove the duplicates using fn.:  df.drop_duplicates()
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [19]:
#check dimension again, should be -5
df_prods_clean_no_dups.shape

(49672, 5)

## Exporting Prepared data -- products_checked.csv

In [20]:
# Export to “Prepared Data” folder.
df_prods_clean_no_dups.to_csv(os.path.join(path,'02 Data','Prepared Data','products_checked.csv'))

# Task 4.5

## Question 1: 
### Perform the consistency checks covered in this Exercise on your df_prods.

In [21]:
for col in df_prods.columns.tolist():
  mix_prods = (df_prods[[col]].applymap(type) != df_prods[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_prods[mix_prods]) > 0:
    print (col)

product_name


In [22]:
df_prods.head()

product_id                                       product_name  aisle_id  \
0           1                         Chocolate Sandwich Cookies        61   
1           2                                   All-Seasons Salt       104   
2           3               Robust Golden Unsweetened Oolong Tea        94   
3           4  Smart Ones Classic Favorites Mini Rigatoni Wit...        38   
4           5                          Green Chile Anytime Sauce         5   

   department_id  prices  
0             19     5.8  
1             13     9.3  
2              7     4.5  
3              1    10.5  
4             13     4.3

In [23]:
# Change column "product_name" to string
df_prods['product_name'] = df_prods['product_name'].astype('str')

## Question 2: 
### Calculate descriptive statistics on df_prods. Share whether anything about the data looks off or should be investigated further.

In [24]:
df_prods.describe()

product_id      aisle_id  department_id        prices
count  49693.000000  49693.000000   49693.000000  49693.000000
mean   24844.345139     67.770249      11.728433      9.994136
std    14343.717401     38.316774       5.850282    453.519686
min        1.000000      1.000000       1.000000      1.000000
25%    12423.000000     35.000000       7.000000      4.100000
50%    24845.000000     69.000000      13.000000      7.100000
75%    37265.000000    100.000000      17.000000     11.200000
max    49688.000000    134.000000      21.000000  99999.000000

### Everything looks fine to me, the only thing that I would like to know is which item(s) cost 99,999dlls.  This could be an error or could be in fact a really expensive item at the grocery store!

## Question 3: 
### Check for mixed-type data in your df_ords dataframe

In [25]:
for col in df_ords.columns.tolist():
  mix_data = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[mix_data]) > 0:
    print (col)

## Question 4: 
### If found mixed data type, fix it

### Did not get anything back, so I can assume that none of the columns have mixed types. 

## Question 5: 
### Check for missing values in your df_ords dataframe

In [26]:
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_time                     0
days_since_prior_order    206209
dtype: int64

### There are 206,209 missing values on the column "days_since_prior_order". This is expected because every single new customer that enters an order for the first time, has never ordered before. We cannot impute a zero or just get rid of them because these are new customers. 

## Question 6: 
### Address the missing values using an appropriate method.

In [27]:
df_ords_nan = df_ords[df_ords['days_since_prior_order'].isnull() == True]
df_ords_nan

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0                 0   2539329        1             1                   2   
11               11   2168274        2             1                   2   
26               26   1374495        3             1                   1   
39               39   3343014        4             1                   6   
45               45   2717275        5             1                   3   
...             ...       ...      ...           ...                 ...   
3420930     3420930    969311   206205             1                   4   
3420934     3420934   3189322   206206             1                   3   
3421002     3421002   2166133   206207             1                   6   
3421019     3421019   2227043   206208             1                   1   
3421069     3421069   3154581   206209             1                   3   

         order_time  days_since_prior_order  
0                 8                     NaN  
11               11                     NaN  
26               14                     NaN  
39               11                     NaN  
45               12                     NaN  
...             ...                     ...  
3420930          12                     NaN  
3420934          18                     NaN  
3421002          19                     NaN  
3421019          15                     NaN  
3421069          11                     NaN  

[206209 rows x 7 columns]

In [28]:
# So we can see that all of these missing values under "Days_since_prior_order" are in fact new customers ordering for the first time. 
# We can create a new column indicating this is the first order. 
# defining a new data frame for cleaned data
df_ords_new = df_ords
df_ords_new['first_order'] = df_ords['days_since_prior_order'].isnull() == True

In [29]:
df_ords_new.head()
# To check the new column

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0           0   2539329        1             1                   2   
1           1   2398795        1             2                   3   
2           2    473747        1             3                   3   
3           3   2254736        1             4                   4   
4           4    431534        1             5                   4   

   order_time  days_since_prior_order  first_order  
0           8                     NaN         True  
1           7                    15.0        False  
2          12                    21.0        False  
3           7                    29.0        False  
4          15                    28.0        False

## Question 7: 
### Run a check for duplicate values in your df_ords data

In [30]:
df_ords_dups = df_ords_new[df_ords_new.duplicated()]
df_ords_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, orders_day_of_week, order_time, days_since_prior_order, first_order]
Index: []

## Question 8: 
### Address the duplicates using an appropriate method.

### There are no duplicates!

## Question 9: 
### Export your final, cleaned df_prods and df_ords data as “.csv” files in your “Prepared Data” folder 

In [31]:
df_ords_new.to_csv(os.path.join(path, '02 Data','Prepared Data', 'orders_checked.csv'))
## I exported products_checked at the end of the lesson, or before question #1